crates/volta-migrate/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+//! Provides types for modeling the current state of the Volta directory and for migrating between versions
+//!
+//! A new layout should be represented by it's own struct (as in the existing v0 or v1 modules)
+//! Migrations between types should be represented by `TryFrom` implementations between the layout types
+//! (see v1.rs for examples)
+//!
+//! NOTE: Since the layout file is written once the migration is complete, all migration implementations
```

Comment:
So IIUC you're saying migrations are supposed to be transactional? I feel like that would mean they should have to write into a temp directory before completing, in case of failure partway through? But I don't see that happening. Am I misunderstanding?

Reply:
My thought was less that they need to be transactional, and more that the migration method should be idempotent, so if it's interrupted we can try again without it erroring out due to being partially applied.

I'm happy to have a deeper discussion about the specific mechanics of migrations going forward, as this was definitely something I struggled with even while making this PR. My biggest concern with something like a temp directory is it would mean we need to copy the entire config directory, apply the changes to it, and then finally remove and replace the existing directory, which is a _lot_ more filesystem operations (especially on Windows / NTFS) than the current model.